mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorUsage.WEBGPU_INTEROP` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorUsage.READ_FROM` and `MLTensorUsage.WRITE_TO` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
+
+### Importing an `MLTensor` to WebGPU
+
+Any `MLTensor` created with the `MLTensorUsage.WEBGPU_INTEROP` flag may be imported into any `GPUDevice`, though cross-device buffer sharing may require expensive data copies. Sharing the tensor requires coordinating between the respective WebNN and WebGPU timelines. Below is an example of how the user agent may coordinate this handoff:
+
+- Two fences are created:
+  1. a "start access" fence which is to be signaled by WebNN and waited on by WebGPU. A data copy may be required alongside the signaling of this fence
+  2. an "end access" fence which is to be signaled by WebGPU and waited on by WebNN. A data copy may be required alongside the signaling of this fence
+- The `GPUDevice` enqueues a command to its `GPUQueue` to wait for the "start access" fence to be signaled
+- WebNN will signal the "start access" fence after the completion of all currently-enqueued operations that use the `MLTensor` which is to be imported (this is very similar to how [`GPUBuffer.mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) works)
+- Until the "end access" fence is signaled:
+  - The `GPUDevice` has exclusive, read/write access to the imported buffer
+  - All WebNN work involving the imported `MLTensor` is blocked
+- When the `GPUBuffer` is destroyed, the "end access" fence is signaled and the `MLTensor` may be used again by WebNN
+
+### `compute()` vs. `dispatch()`
+
+`compute()` will be deprecated and removed in favor of `dispatch()`.
+
+It's possible `compute()` may have a performance advantage on some platforms for some use cases, such as in one-shot inference (where the benefits of buffer re-use are not relevant) with small inputs/outputs on CPU (where the overhead of task queueing may outweigh the benefits of parallelization). At this time, we do not expect this performance impact to be substantial enough to justify providing two mostly-identical methods for executing an `MLGraph`.
+
+### Open Questions
+
+- How will errors be surfaced? Do we need a concept similar to [WebGPU's error scopes](https://www.w3.org/TR/webgpu/#error-scopes), or is [returning errors via a promise for select operations sufficient](https://github.com/webmachinelearning/webnn/issues/697#issuecomment-2195656878)? See [#477](https://github.com/webmachinelearning/webnn/issues/477)
```

Comment:
WebGPU has error scopes because the API is very stateful and it's non-trivial to replicate all of the state checking in both the JS process and the GPU process. ErrorScopes can be challenging for web developers to use. 

For WebNN, if implementations can efficiently check state in the JS process and throw exceptions on errors, we shouldn't need error scopes. For security, of course, the same state checking will also need to happen in the GPU process since a p000ned JS process can bypass error checking and send the GPU process whatever it wants. 

In summary, in think we can leave ErrorScopes out of the spec for now and revisit if needed.